### PR TITLE
std: Stabilize the `hash` module

### LINKED ISSUE #22467: Tracking issue for Simplify `std::hash` (RFC 823)
https://github.com/rust-lang/rfcs/pull/823

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is an implementation of [RFC 823](https://github.com/rust-lang/rfcs/blob/master/text/0823-hash-simplification.md) which is another pass over
the `std::hash` module for stabilization. The contents of the module were not
entirely marked stable, but some portions which remained quite similar to the
previous incarnation are now marked `#[stable]`. Specifically:
- `std::hash` is now stable (the name)
- `Hash` is now stable
- `Hash::hash` is now stable
- `Hasher` is now stable
- `SipHasher` is now stable
- `SipHasher::new` and `new_with_keys` are now stable
- `Hasher for SipHasher` is now stable
- Many `Hash` implementations are now stable

All other portions of the `hash` module remain `#[unstable]` as they are less
commonly used and were recently redesigned.

This commit is a breaking change due to the modifications to the `std::hash` API
and more details can be found on the [RFC](https://github.com/rust-lang/rfcs/blob/master/text/0823-hash-simplification.md).

Closes #22467
[breaking-change]

### CODE REVIEW

**Comment by alexcrichton** [comment]
The diff stat is a bit alarming, but it's 99% due to `#[cfg(stage0)]` which is in turn due to `derive(Hash)`.

File: src/libcore/hash/mod.rs [hunk 1]
```diff
@@ -76,22 +72,123 @@ mod sip;
 /// to compute the hash. Specific implementations of this trait may specialize
 /// for particular instances of `H` in order to be able to optimize the hashing
 /// behavior.
+#[cfg(not(stage0))]
```
Comment (aturon) [inline src/libcore/hash/mod.rs]: The doc comment above here is slightly stale -- some of it should move to the methods, I think.

**Comment by aturon** [comment]
So... I don't know how to meaningfully review the 4+kloc of stage0 fallout here, but the changes to the core module LGTM. Let me know if there's any particular area of fallout that you think wants more scrutiny.

**Pull request merged**